typing/uniqueness_analysis.ml

```diff
@@ -0,0 +1,1302 @@
+     it was the result of a single path: using a parent shared and a child
+     uniquely is obviously bad. Howerver, it might be the result of "par"ing
+     multiple path: `S` par `N -> U`, which is valid. *)
+
+  (* INVARIANT: children >= parent. For example, having a shared child under a
+     unique parent is nonsense. The invariant is preserved because Usage.par and
+     Usage.seq above is monotone, and UsageTree.par and UsageTree.seq here is
+     node-wise. *)
+  type t = { children : t Projection.Map.t; usage : Usage.t }
+
+  let rec print_children ppf children =
+    Projection.Map.iter
+      (fun proj child ->
+        Format.fprintf ppf "%a = %a," Projection.print proj print child)
+      children
+
+  and print ppf t =
+    Format.fprintf ppf "%a {%a}" Usage.print t.usage print_children t.children
+
+  module Path = struct
+    type t = Projection.t list
+
+    let child (p : t) (a : Projection.t) : t = p @ [ a ]
+
+    (** length of path ignoring memory_address *)
+    let rec real_length = function
+      | [] -> 0
+      | Projection.Memory_address :: xs -> real_length xs
+      | _ :: xs -> real_length xs + 1
+
+    let root : t = []
+    let _print ppf = Format.pp_print_list Projection.print ppf
+
+    let _human_readable ppf t =
+      Format.pp_print_list
+        ~pp_sep:(fun ppf () -> Format.fprintf ppf " which is ")
+        Projection.human_readable ppf t
+  end
+
+  exception
+    MultiUse of {
+      (* the occurrence that's failing forcing *)
+      here : Occurrence.t;
+      (* the other occurrence for reference *)
+      there : Occurrence.t;
+      (* which axis failed to force? *)
+      error : SharedUnique.error;
+      (* descentdant means there is a descendant of here *)
+      there_is_of_here : relation;
+      (* path from the ancestor to the descendant, reversed *)
+      path : Path.t;
+    }
+
+  let leaf usage = { usage; children = Projection.Map.empty }
+  let empty = leaf Usage.Unused
+
+  (* lift par and seq to trees *)
+
+  let rec par t0 t1 =
+    {
+      usage = Usage.par t0.usage t1.usage;
+      children =
+        Projection.Map.merge
+          (fun _proj c0 c1 ->
+            let c0 = Option.value c0 ~default:t0 in
+            let c1 = Option.value c1 ~default:t1 in
+            Some (par c0 c1))
+          t0.children t1.children;
+    }
+
+  (* projs0 is the list of projections for t0.
+     It is from the ancestor that t0.usage actually comes from,
+     to t0.*)
+  let rec seq t0 projs0 t1 projs1 =
+    let usage =
+      try Usage.seq t0.usage t1.usage
+      with SharedUnique.MultiUse { here; there; error } ->
+        (* t' is ancestor *)
+        let t', path =
+          match (projs0, projs1) with
+          | [], _ -> (t1.usage, projs1)
+          | _, [] -> (t0.usage, projs0)
+          | _, _ -> assert false
+        in
+        let there_is_of_here =
+          if (Usage.extract_occurrence t').loc = here.loc then Descendant
```

Comment:
I was on the extreme side of modularization. There are two `MultiUse` exception, one in `Usage` and the other in `UsageTree`. The latter is the former lifted to the tree structure and thus contains the relative path between the two conflicting occurrences. The code you are refering to is deciding the ancestor-descendant relation between the two.

This code is now removed, since we no longer print the path between the two occurrences.